### PR TITLE
Implement first steps of the dual funding flow

### PR DESCRIPTION
**Pull request by t-bast**
We implement the first step of the dual funding protocol: exchanging `open_channel2` and `accept_channel2`.
We currently stop after exchanging these two messages.
Future commits will add the `interactive-tx` protocol used to build the funding transaction.

NB: this builds on top of #2245

### CODE REVIEW

**Review by pm47 — COMMENTED** [review]
Just a first thought, I'll review the rest tomorrow

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala [hunk 1]
```diff
@@ -208,43 +209,77 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder, val
...
-          ChannelTlv.UpfrontShutdownScriptTlv(localShutdownScript),
-          ChannelTlv.ChannelTypeTlv(channelType)
-        ))
-      goto(WAIT_FOR_ACCEPT_CHANNEL) using DATA_WAIT_FOR_ACCEPT_CHANNEL(initFunder, open) sending open
-
-    case Event(inputFundee@INPUT_INIT_FUNDEE(_, localParams, remote, _, _, _), Nothing) if !localParams.isInitiator =>
-      activeConnection = remote
-      txPublisher ! SetChannelId(remoteNodeId, inputFundee.temporaryChannelId)
-      goto(WAIT_FOR_OPEN_CHANNEL) using DATA_WAIT_FOR_OPEN_CHANNEL(inputFundee)
+    case Event(input: INPUT_INIT_CHANNEL_INITIATOR, Nothing) =>
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:212]: True, that could make sense, anything we can do to make the `Channel.scala` file slimmer would be useful. I'll try that.

[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:211]: Done in https://github.com/ACINQ/eclair/pull/2247/commits/0adf29ea637ed46240d6a2832f8c53c3785a07b5

I didn't introduce separate case classes, as it created a lot of unnecessary duplication as they're very similar, and force more switching in the `Peer` where we would create them. I can still add it though if you feel it's worth the extra code.

[the remark this answers is not in the document]
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:211]: That's fine

The symmetry feels right 😎 

https://github.com/ACINQ/eclair/blob/0adf29ea637ed46240d6a2832f8c53c3785a07b5/eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala#L217-L221

https://github.com/ACINQ/eclair/blob/0adf29ea637ed46240d6a2832f8c53c3785a07b5/eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala#L226-L230

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 2]
```diff
@@ -201,6 +237,35 @@ object Helpers {
...
+    extractShutdownScript(accept.temporaryChannelId, localFeatures, remoteFeatures, accept.upfrontShutdownScript_opt).map(script_opt => (channelFeatures, script_opt))
+  }
+
+  /** Compute the channelId of a dual-funded channel. */
+  def computeChannelId(open: OpenDualFundedChannel, accept: AcceptDualFundedChannel): ByteVector32 = {
+    if (LexicographicalOrdering.isLessThan(open.revocationBasepoint.value, accept.revocationBasepoint.value)) {
+      Crypto.sha256(open.revocationBasepoint.value ++ accept.revocationBasepoint.value)
+    } else {
+      Crypto.sha256(accept.revocationBasepoint.value ++ open.revocationBasepoint.value)
+    }
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:266]: Alternatively:
```suggestion
    val bin = Seq(open.revocationBasepoint.value, accept.revocationBasepoint.value)
      .sortWith(LexicographicalOrdering.isLessThan) // sort
      .reduce(_ ++ _) // concatenate
    Crypto.sha256(bin)
```

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 3]
```diff
@@ -134,61 +134,45 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnChainA
...
-          val fundingTxFeeratePerKw = c.fundingTxFeeratePerKw_opt.getOrElse(nodeParams.onChainFeeConf.feeEstimator.getFeeratePerKw(target = nodeParams.onChainFeeConf.feeTargets.fundingBlockTarget))
-          log.info(s"requesting a new channel with type=$channelType fundingSatoshis=${c.fundingSatoshis}, pushMsat=${c.pushMsat} and fundingFeeratePerByte=${c.fundingTxFeeratePerKw_opt} temporaryChannelId=$temporaryChannelId localParams=$localParams")
-          channel ! INPUT_INIT_FUNDER(temporaryChannelId, c.fundingSatoshis, c.pushMsat, channelFeeratePerKw, fundingTxFeeratePerKw, localParams, d.peerConnection, d.remoteInit, c.channelFlags.getOrElse(nodeParams.channelConf.channelFlags), channelConfig, channelType)
+          val temporaryChannelId = if (dualFunded) {
+            val channelKeyPath = nodeParams.channelKeyManager.keyPath(localParams, channelConfig)
+            val revocationBasepoint = nodeParams.channelKeyManager.revocationPoint(channelKeyPath).publicKey
+            Crypto.sha256(ByteVector.fill(33)(0) ++ revocationBasepoint.value)
+          } else {
+            randomBytes32()
+          }
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:159]: Probably belong in `Helpers`?

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 4]
```diff
@@ -379,6 +363,51 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnChainA
...
+            log.info(s"accepting a new channel with type=$channelType temporaryChannelId=$temporaryChannelId localParams=$localParams")
+            msg match {
+              case Left(openSingleFunder) =>
+                channel ! INPUT_INIT_CHANNEL_NON_INITIATOR(temporaryChannelId, None, dualFunded = false, localParams, d.peerConnection, d.remoteInit, channelConfig, channelType)
+                channel ! openSingleFunder
+              case Right(openDualFunded) =>
+                // NB: we don't add a contribution to the funding amount.
+                channel ! INPUT_INIT_CHANNEL_NON_INITIATOR(temporaryChannelId, None, dualFunded = true, localParams, d.peerConnection, d.remoteInit, channelConfig, channelType)
+                channel ! openDualFunded
+            }
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:397]: This `match/case` and the number of `fold` in the method make me think `acceptOrRejectChannel` is not the right factorization.

Maybe just factor the channel type negotiation?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelData.scala [hunk 5]
```diff
@@ -47,13 +47,21 @@ import java.util.UUID
  */
 sealed trait ChannelState
 case object WAIT_FOR_INIT_INTERNAL extends ChannelState
+// Single-funder channel opening:
+case object WAIT_FOR_INIT_CHANNEL extends ChannelState
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelData.scala:51]: Since none of the other states for single-funder explicitly mention `SINGLE_FUNDED`, I wanted to stay consistent and not mention it, but we can!

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelFeatures.scala [hunk 6]
```diff
@@ -59,7 +59,7 @@ object ChannelFeatures {
   def apply(channelType: ChannelType, localFeatures: Features[InitFeature], remoteFeatures: Features[InitFeature]): ChannelFeatures = {
     // NB: we don't include features that can be safely activated/deactivated without impacting the channel's operation,
     // such as option_dataloss_protect or option_shutdown_anysegwit.
-    val availableFeatures = Seq(Features.Wumbo, Features.UpfrontShutdownScript).filter(f => Features.canUseFeature(localFeatures, remoteFeatures, f))
+    val availableFeatures = Seq(Features.Wumbo, Features.UpfrontShutdownScript, Features.DualFunding).filter(f => Features.canUseFeature(localFeatures, remoteFeatures, f))
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelFeatures.scala:62]: Good idea, I'll try that.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 7]
```diff
@@ -153,22 +151,60 @@ object Helpers {
...
   }
 
-  /**
-   * Called by the funder
-   */
-  def validateParamsFunder(nodeParams: NodeParams, channelType: SupportedChannelType, localFeatures: Features[InitFeature], remoteFeatures: Features[InitFeature], open: OpenChannel, accept: AcceptChannel): Either[ChannelException, (ChannelFeatures, Option[ByteVector])] = {
-    accept.channelType_opt match {
-      case Some(theirChannelType) if accept.channelType_opt != open.channelType_opt =>
+  /** Called by the non-initiator of a dual-funded channel. */
+  def validateParamsNonInitiator(nodeParams: NodeParams, channelType: SupportedChannelType, open: OpenDualFundedChannel, remoteNodeId: PublicKey, localFeatures: Features[InitFeature], remoteFeatures: Features[InitFeature]): Either[ChannelException, (ChannelFeatures, Option[ByteVector])] = {
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:155]: No, the requirements for dual-funding are different from single-funder, there are single-funder requirements that don't make sense anymore for dual-funding, so this should be reviewed by comparing it to the requirements of the single-funder case.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala (hunk 4 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:397]: Good idea, I'll try that.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelFeatures.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelFeatures.scala:62]: This is quite orthogonal, so I'll do that in a separate PR today to ensure it's properly reviewed on its own.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 2 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala]: Done in https://github.com/ACINQ/eclair/pull/2247/commits/ab71773b7349191602c43e6b3e175fab8d06f159

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala (hunk 3 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala]: Done in https://github.com/ACINQ/eclair/pull/2247/commits/ab71773b7349191602c43e6b3e175fab8d06f159

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala (hunk 4 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala]: Done in https://github.com/ACINQ/eclair/pull/2247/commits/3a0d2356c42ca22bee145deacb3c277848be0020, the remaining small duplication between the two handlers is completely acceptable imo.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelFeatures.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelFeatures.scala:62]: Done in #2282, there is a subtlety (this new trait must exclude channel type features), let's discuss on that PR whether this is the right way to do it or not

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 8]
```diff
@@ -201,6 +237,41 @@ object Helpers {
...
   }
 
+  /** Called by the initiator of a dual-funded channel. */
+  def validateParamsInitiator(nodeParams: NodeParams, channelType: SupportedChannelType, localFeatures: Features[InitFeature], remoteFeatures: Features[InitFeature], open: OpenDualFundedChannel, accept: AcceptDualFundedChannel): Either[ChannelException, (ChannelFeatures, Option[ByteVector])] = {
+    validateChannelType(open.temporaryChannelId, channelType, open.channelType_opt, accept.channelType_opt, localFeatures, remoteFeatures) match {
+      case Some(t) => return Left(t)
+      case None => // we agree on channel type
+    }
+
+    if (accept.maxAcceptedHtlcs > Channel.MAX_ACCEPTED_HTLCS) return Left(InvalidMaxAcceptedHtlcs(accept.temporaryChannelId, accept.maxAcceptedHtlcs, Channel.MAX_ACCEPTED_HTLCS))
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:247]: Why do we need it?

[the remark this answers is not in the document]
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:247]: Ha! good question. @sstone any idea? As always it's mostly the inconsistency with the single funder case which bothers me.

[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:247]: We can just remove it from the single funder case :)
Maybe that was used to test edge cases on testnet a very very long time ago?
I really don't see why it would be useful today.

**Comment by t-bast** [comment]
Rebased to fix conflicts with #2286

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:247]: It seems that in bitcoin core dust is handled the same on mainnet and testnet so I don't see why we needed it ? maybe introp issues with cln or lnd a long time ago ? anyway I think it is safe to have the same checks on mainnet and testnet.

[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:223]: Removed in the previous handlers as well in https://github.com/ACINQ/eclair/pull/2247/commits/2e766f4b1e511f8b341309e6951d2b6fa96ca8ef

**Comment by t-bast** [comment]
Rebased to fix conflicts with #2224, I had to squash commits otherwise the rebase was too complex/risky.

**Comment by pm47** [comment]
ℹ️ Needs rebase.

**Comment by t-bast** [comment]
Rebased.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 9]
```diff
@@ -141,24 +141,24 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnChainA
...
+          val channelFlags = c.channelFlags_opt.getOrElse(nodeParams.channelConf.channelFlags)
+          val channelType = c.channelType_opt.getOrElse(ChannelTypes.defaultFromFeatures(d.localFeatures, d.remoteFeatures, channelFlags.announceChannel))
           // NB: we need to capture parameters in a val to use them in andThen
           val selfRef = self
           val origin = sender()
           implicit val ec: ExecutionContext = ExecutionContext.Implicits.global
-          createLocalParams(nodeParams, d.localFeatures, channelType, isInitiator = true, c.fundingSatoshis).andThen {
+          createLocalParams(nodeParams, d.localFeatures, channelType, isInitiator = true, c.fundingAmount).andThen {
             case Success(localParams) => selfRef ! SpawnChannelInitiator(c, ChannelConfig.standard, channelType, localParams, origin)
             case Failure(t) => origin ! Status.Failure(new RuntimeException("channel creation failed", t))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:163]: Nit (to preserve sender):
```suggestion
            case Failure(t) => origin.tell(Status.Failure(new RuntimeException("channel creation failed", t)), selfRef)
```

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 10]
```diff
@@ -168,54 +168,83 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnChainA
...
+        }
+        val fundingTxFeerate = c.fundingTxFeerate_opt.getOrElse(nodeParams.onChainFeeConf.feeEstimator.getFeeratePerKw(target = nodeParams.onChainFeeConf.feeTargets.fundingBlockTarget))
+        val commitTxFeerate = nodeParams.onChainFeeConf.getCommitmentFeerate(remoteNodeId, channelType, c.fundingAmount, None)
+        log.info(s"requesting a new channel with type=$channelType fundingAmount=${c.fundingAmount} dualFunded=$dualFunded pushAmount=${c.pushAmount_opt} fundingFeerate=$fundingTxFeerate temporaryChannelId=$temporaryChannelId localParams=$localParams")
+        channel ! INPUT_INIT_CHANNEL_INITIATOR(temporaryChannelId, c.fundingAmount, dualFunded, commitTxFeerate, fundingTxFeerate, c.pushAmount_opt, localParams, d.peerConnection, d.remoteInit, c.channelFlags_opt.getOrElse(nodeParams.channelConf.channelFlags), channelConfig, channelType)
         stay() using d.copy(channels = d.channels + (TemporaryChannelId(temporaryChannelId) -> channel))
 
-      case Event(msg: protocol.OpenChannel, d: ConnectedData) =>
-        d.channels.get(TemporaryChannelId(msg.temporaryChannelId)) match {
+      case Event(open: protocol.OpenChannel, d: ConnectedData) =>
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:183]: Handlers for `protocol.OpenChannel` and `protocol.OpenDualFundedChannel` could be factored.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala [hunk 11]
```diff
@@ -208,43 +208,26 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder, val
...
+      context.system.eventStream.publish(ChannelCreated(self, peer, remoteNodeId, isInitiator = true, input.temporaryChannelId, input.commitTxFeerate, Some(input.fundingTxFeerate)))
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:212]: I noticed that as well, but it is previous behavior that is unrelated to this PR, so I chose to ignore it for now.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala (hunk 10 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:183]: That's what I started with, but then we reverted it after this comment: https://github.com/ACINQ/eclair/pull/2247#discussion_r878212163

There are similarities, but there are enough small differences to make it messy to completely factor together...I'll give it another try though.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelData.scala [hunk 12]
```diff
@@ -47,13 +47,21 @@ import java.util.UUID
  */
 sealed trait ChannelState
 case object WAIT_FOR_INIT_INTERNAL extends ChannelState
+// Single-funder channel opening:
+case object WAIT_FOR_INIT_SINGLE_FUNDER_CHANNEL extends ChannelState
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelData.scala]: Done in https://github.com/ACINQ/eclair/pull/2247/commits/88260420dab107cda02d1569b9712f72f16d1e64

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala (hunk 10 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:183]: I have a better middle-ground in https://github.com/ACINQ/eclair/pull/2247/commits/31c3ee911cca907b62b82f5bc8ecedcaa1ad0563, let me know what you think

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala (hunk 9 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala]: Done in https://github.com/ACINQ/eclair/pull/2247/commits/31c3ee911cca907b62b82f5bc8ecedcaa1ad0563

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala (hunk 10 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:183]: I think it's good enough, we can always revisit later.

**Review by pm47 — DISMISSED** [review]
LGTM!
